File: JavaScript/2196_Create_Binary_Tree_From_Descriptions/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[][]} descriptions
 * @return {TreeNode}
 */
const createBinaryTree = function (descriptions) {
    const nodeMap = new Map();
    const children = new Set();

    descriptions.forEach((val) => {
        children.add(val[1]);

        if (!nodeMap.has(val[0])) {
            nodeMap.set(val[0], new TreeNode(val[0]));
        }
        if (!nodeMap.has(val[1])) {
            nodeMap.set(val[1], new TreeNode(val[1]));
        }

        const parentNode = nodeMap.get(val[0]);
        const childNode = nodeMap.get(val[1]);

        if (val[2]) {
            parentNode.left = childNode;
        } else {
            parentNode.right = childNode;
        }
    });

    for (let val of descriptions) {
        if (!children.has(val[0])) {
            return nodeMap.get(val[0]);
        }
    }
};

File: Go/2196_Create_Binary_Tree_From_Descriptions/solution.go
package _2196_Create_Binary_Tree_From_Descriptions

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func createBinaryTree(descriptions [][]int) *TreeNode {
	nodeMap := make(map[int]*TreeNode)
	children := make(map[int]bool)

	for _, val := range descriptions {
		children[val[1]] = true

		if nodeMap[val[0]] == nil {
			nodeMap[val[0]] = new(TreeNode)
			nodeMap[val[0]].Val = val[0]
		}
		if nodeMap[val[1]] == nil {
			nodeMap[val[1]] = new(TreeNode)
			nodeMap[val[1]].Val = val[1]
		}

		if val[2] == 1 {
			nodeMap[val[0]].Left = nodeMap[val[1]]
		} else {
			nodeMap[val[0]].Right = nodeMap[val[1]]
		}
	}

	var res *TreeNode
	for _, val := range descriptions {
		if children[val[0]] != true {
			res = nodeMap[val[0]]
		}
	}

	return res
}
